### ISSUE TITLE
asyncssh server on an already connected socket

### ISSUE THREAD

**Issue by silentnights** [ronf/asyncssh#205]
Hi,

I wonder if it is possible to start an asyncssh server on an already connected socket? This is kind of like doing a reverse ssh connection, but just on a simple tcp socket.

I have already implemented something that worked. Basically I copied the create_server from asyncssh/connection.py and changed it to return SSHServerConnection, and used this as a protocol_factory for asyncio.create_connection, which I pass the already connected socket to it.

But even that it works, I still do all the internal initialization that create_server do in my protocol_factory which calls some asyncio internal methods and use asyncio internal variables. So it is not an ideal solution.

So is there a better way to do it?

**Comment by ronf** [comment]
You mentioned create_connection() here -- does that means you are trying to make an outbound TCP connection but then use that connection to run SSH "in reverse", with the thing you are connecting to acting as the client and you acting as the server, or were you just looking to take an inbound TCP connection your code has already done a TCP accept() on and run an SSH server on that?

**Comment by ronf** [comment]
Actually, I see that asyncio.create_connection() can also take an already connected socket as an argument, so you could potentially pass either an inbound or outbound socket to that, if you don't need it to do the connect for you. I'm still curious which of the cases you were thinking of, though.

I added something close to this on the client side already, in the form of a "tunnel=" argument that is designed to take an SSHClientConnection for tunneling SSH within an existing SSH client connection, but the tunnel argument can also be an instance of your own class with a create_connection() method that returns a transport and protocol pair. So, you could use it with BaseEventLoop.create_connection() where you pass in an existing socket. I've thought about extending "tunnel=" to just directly take a socket object as well, as a shortcut for that case, or perhaps just adding a separate "sock" argument for that, similar to what exists on BaseEventLoop.create_connection(). However, you'd still need to pass in arguments like the destination host & port to allow things like the known_hosts lookup to work, so that didn't quite feel right.

I don't have anything like this on the server side yet, but I'm open to exploring the idea of the best way to fit that in. As you noted, the existing create_server() code has pretty much everything needed, and so that code could be factored out into a common function that could be used for both create_server() and a one-shot call that ran on an existing connected socket.

**Comment by ronf** [comment]
Looking more closely, it'd be difficult to avoid a copy/paste of a lot of the existing create_server() definition since I'd have to a large number of the arguments, but one thought would be to add a top-level asyncssh.accept_connection() function that accepts a "sock" argument instead of "host" and "port" and drops the related arguments like "family", "flags", "backlog", "reuse_address", and "reuse_port". It could then return an SSHServerConnection and associated owner (typically a subclass of SSHServer) similar to what is returned by asyncssh.create_connection() on the client side.

One interesting question: Would you want this new call to delay until after the authentication of the client was complete like create_connection() does on the client side or return immediately, before the SSH handshake and authentication were done, operating in the background more like create_server() even though this would only be for a single connection?

**Comment by silentnights** [comment]
> es that means you are trying to make an outbound TCP connection but then use that connection to run SSH "in reverse", with the thing you are connecting to acting as the client and you acting as the server, or were you just looking to take an inbound TCP connection your code has already done a TCP accept() on and run an SSH server on t

Yes I am trying to establish a TCP connection to the client and then the client tries to establish an SSH session on that socket while me acting as an SSH server. I am doing this to implement https://tools.ietf.org/html/rfc8071

**Comment by silentnights** [comment]
> Looking more closely, it'd be difficult to avoid a copy/paste of a lot of the existing create_server() definition since I'd have to a large number of the arguments, but one thought would be to add a top-level asyncssh.accept_connection() function that accepts a "sock" argument instead of "host" and "port" and drops the related arguments like "family", "flags", "backlog", "reuse_address", and "reuse_port". It could then return an SSHServerConnection and associated owner (typically a subclass of SSHServer) similar to what is returned by asyncssh.create_connection() on the client side.
> 
> One interesting question: Would you want this new call to delay until after the authentication of the client was complete like create_connection() does on the client side or return immediately, before the SSH handshake and authentication were done, operating in the background more like create_server() even though this would only be for a single connection?

hmm. I think delaying until the authentication of the client was complete is useful since I need a single SSH connection at the end. Specially that in case the connection failed to authenticate, maybe I will need to do something else, this would be different from create_server since even if a single connection fails to authenticate, it will still accept more connection in the background.

But still it is good to think if returning directly would be useful for other cases? I am new to asyncssh so I am not sure if you return directly like in create_server() would you still be able to figure out if the SSH connection was successful or not?

I am actually thinking of something different, One of the problems I had is that create_server takes a lot of optional parameters, and SSHServerConnection takes the parameters without defaults so create_server has to prepare the parameters for SSHServerConnection.

So if we can either accept optional parameters in SSHServerConnection which establish the correct parameters like what you do now in create_server, or create something like DefaultSSHServerConnection which inherits from SSHServerConnection but initialize the parameters of it's parent class, then someone can use that DefaultSSHServerConnection and pass it as a protocol_factory for create_connection. What do you think?

**Comment by ronf** [comment]
> hmm. I think delaying until the authentication of the client was complete is useful since I need a single SSH connection at the end. Specially that in case the connection failed to authenticate, maybe I will need to do something else, this would be different from create_server since even if a single connection fails to authenticate, it will still accept more connection in the background.

Yeah - that's what I was thinking as well. Waiting would also allow any exception which occurred during the handshake and authentication to be raised directly on the call which "accepts" the connection as a new SSHServerConnection.

> But still it is good to think if returning directly would be useful for other cases? I am new to asyncssh so I am not sure if you return directly like in create_server() would you still be able to figure out if the SSH connection was successful or not?

If it returned immediately, it would still be able to return an SSHServerConnection and the connection "owner". This would probably happen right after calling connection_made() on the new SSHServerConnection. In this case, any errors would be reported through a call to connection_lost() on the owner (created by a call to the server_factory passed in to the accept). It all comes down to whether we have this act more like create_connection() or create_server() in terms of error handling. I'm torn on this.

> I am actually thinking of something different, One of the problems I had is that create_server takes a lot of optional parameters, and SSHServerConnection takes the parameters without defaults so create_server has to prepare the parameters for SSHServerConnection.
> 
> So if we can either accept optional parameters in SSHServerConnection which establish the correct parameters like what you do now in create_server, or create something like DefaultSSHServerConnection which inherits from SSHServerConnection but initialize the parameters of it's parent class, then someone can use that DefaultSSHServerConnection and pass it as a protocol_factory for create_connection. What do you think?

I definitely see the value in having something like this. It wasn't all that necessary with create_server(), as the preparing of the SSHServerConnection arguments could be done once and shared for all connections, but with the new "accept" call, that would be more difficult to do. I don't really like the idea of the caller subclassing SSHServerConnection to achieve this, though.

I think what makes sense here would be to create a new class called something like SSHServerConnectionOptions which takes all of the keyword arguments to create_server() today that end up getting passed to SSHServerConnection. All of the work to prepare these arguments could be moved to the constructor of this new class, and the existing create_server() call could be changed to just pass all of these existing keyword arguments through to this new class, keeping around the resulting "options" object to pass into SSHServerConnection() as its only argument in the new scheme. The create_server() call could also take an optional "options" argument where you could pass in an SSHServerConnectionOptions object which already had all the arguments prepared -- this would be mutually exclusive with passing any of the current keyword arguments. This same "options" argument could also be passed to the new "accept" method if you wanted to prepare the options only once and use them for multiple "accept" calls. It could also take the keyword arguments directly if you wanted to prepare a fresh set of options each time or only planned on calling "accept" once.

What do you think?

**Comment by ronf** [comment]
Also, thanks for the added context about RFC 8071. I had seen NETCONF over SSH before, but not the "Call Home" variant of it. This is definitely something I'd like to provide a clean solution for.

**Comment by silentnights** [comment]
> If it returned immediately, it would still be able to return an SSHServerConnection and the connection "owner". This would probably happen right after calling connection_made() on the new SSHServerConnection. In this case, any errors would be reported through a call to connection_lost() on the owner (created by a call to the server_factory passed in to the accept). It all comes down to whether we have this act more like create_connection() or create_server() in terms of error handling. I'm torn on this.

The problem here is that it is not clear if someone would need that behavior instead of the other because of some reason. This can not be predicted.

But how about if we allow the user to pass a future as an optional parameter, if the option is passed then create_connection return that future which is set once the authentication is done, this emulates create_connection.
If no future is passed then we can return immediately like in create_server.

That would be flexible enough to cover both scenarios.

> I definitely see the value in having something like this. It wasn't all that necessary with create_server(), as the preparing of the SSHServerConnection arguments could be done once and shared for all connections, but with the new "accept" call, that would be more difficult to do. I don't really like the idea of the caller subclassing SSHServerConnection to achieve this, though.
> 
> I think what makes sense here would be to create a new class called something like SSHServerConnectionOptions which takes all of the keyword arguments to create_server() today that end up getting passed to SSHServerConnection. All of the work to prepare these arguments could be moved to the constructor of this new class, and the existing create_server() call could be changed to just pass all of these existing keyword arguments through to this new class, keeping around the resulting "options" object to pass into SSHServerConnection() as its only argument in the new scheme. The create_server() call could also take an optional "options" argument where you could pass in an SSHServerConnectionOptions object which already had all the arguments prepared -- this would be mutually exclusive with passing any of the current keyword arguments. This same "options" argument could also be passed to the new "accept" method if you wanted to prepare the options only once and use them for multiple "accept" calls. It could also take the keyword arguments directly if you wanted to prepare a fresh set of options each time or only planned on calling "accept" once.
> 
> What do you think?

I really like the SSHServerConnectionOptions object. It is much more flexible that what is available now. You still maintain the parameter validation early enough before creating the server, but also allow someone to have a validated option available which he can keep around. That would be even useful for create_connection if you want to create several connections to several servers with only small subset of the parameters change.

So in general I definitely think we should go for that.

BTW let me know if i can help you with something. I can probably create a patch for this during the weekend.

**Comment by ronf** [comment]
> The problem here is that it is not clear if someone would need that behavior instead of the other because of some reason. This can not be predicted.
> 
> But how about if we allow the user to pass a future as an optional parameter, if the option is passed then create_connection return that future which is set once the authentication is done, this emulates create_connection. If no future is passed then we can return immediately like in create_server.

That might work. I could also just provide a separate call you can make to wait for the auth to complete, though, and it could take care of creating the Future for you and then blocking on it.

In that case, the accept call would always return right away, but if you wanted to wait and see what happened with the auth, you'd block in a "wait" call right after that and it would raise an exception if an error occurred before completing the auth. In fact, I already have something like that in place which gets used to do the client-side wait that would already work for both cases. I just don't document it right now.

> I really like the SSHServerConnectionOptions object. It is much more flexible that what is available now. You still maintain the parameter validation early enough before creating the server, but also allow someone to have a validated option available which he can keep around. That would be even useful for create_connection if you want to create several connections to several servers with only small subset of the parameters change.
> 
> So in general I definitely think we should go for that.

Ok, sounds good. I agree that this is something we could use for creating both client and server connections, and maybe even use in other places like when creating sessions within a connection.

> BTW let me know if i can help you with something. I can probably create a patch for this during the weekend.

I appreciate the offer! I've already got a few ideas about how to make a general "SSHOptions" base class that I can use for this, though. Give me a few days to prototype that and then I'd love to get your feedback on it.

One thing that occurred to me is that I might not need to make this new "options" argument mutually exclusive with other connection setting arguments. I think I can create something here that would let you save a collection of options to reuse but also incrementally change a subset of those saved options by passing in both an options object and keyword args in the connect/accept calls. With no options object, you start from the defaults, changing only the values you specify. With an options object, that essentially replaces the defaults, but you can still incrementally change those settings on a per-connection basis.

I'm also thinking that options objects could be built up from other options objects by passing in an existing options object along with the set of values to change when creating a new one. I might also allow options objects to be updated in place, with any changes taking effect on any future calls that reference that object. All these variants could use the same set of keyword arguments to specify the options, just appearing in different places.

**Comment by ronf** [comment]
A quick update on this: I have implemented a new base class for managing options and created subclasses so far for client connection & server connection options and made all the changes needed to create_connection() and create_server() to use this new capability. I still have a bit of review and cleanup to do, but the new code is passing all the unit tests and the coverage looks good.

I've also been thinking more about how to expose the reverse-direction connection support, and with the new structure I think I may be able to use the existing create_connection() and create_server() commands, but have them take an optional argument like "reverse=True" that would flip around the connection factory used when a new connection was created, as well as the default owner class when that's no specified. So, you could end up in create_connection() using factories for SSHServer/SSHServerConnection once the TCP connection was made or end up in create_server() using factories for SSHClient/SSHClientConnection as each TCP new connection was accepted. I need to do just a bit more refactoring to allow both paths to create both types of objects, but it looks doable. 

If all goes well, I might even be able to make these "reverse" connections work when carried within an SSH tunnel to a jump host, the way AsyncSSH already can today for normal SSH connections.

**Comment by silentnights** [comment]
Thanks a lot for your effort and help.

> I've also been thinking more about how to expose the reverse-direction connection support, and with the new structure I think I may be able to use the existing create_connection() and create_server() commands, but have them take an optional argument like "reverse=True" that would flip around the connection factory used when a new connection was created, as well as the default owner class when that's no specified. So, you could end up in create_connection() using factories for SSHServer/SSHServerConnection once the TCP connection was made or end up in create_server() using factories for SSHClient/SSHClientConnection as each TCP new connection was accepted. I need to do just a bit more refactoring to allow both paths to create both types of objects, but it looks doable.

I think it is probably better to keep the interfaces separate. The reason for this is that the behavior is different. So you would be doing 2 different things using the same interface. Having a flag inside the interface is the exact sign of that, besides theses functions already have a very long list of arguments and I personally think that is confusing specially when you start using asyncssh. Even if you would want to use the very simple case, you can not help it but to wonder what the values of these parameters should be and whether you need to change it or not. That is even more confusing if you are not very experienced with ssh, just like me. :-)

Also note that the use case is special, adding one more flag would confuse the user who would want to use just the simple case for the benefit of the one who would need the very special case.

**Comment by ronf** [comment]
Sorry for the slow response. This turned out to be a bit more involved than I thought to really get this right. After trying out a few different approaches, I concluded that you were right to keep the interfaces separate for the forward & reverse direction. So, I now have connect_reverse() and listen_reverse() methods that look similar to connect() and listen(), but they take the opposite kind of options (so connect_reverse takes server options and listen_reverse takes client options).

I've got the new options classes and the reverse connect/listen functions working, but I still need to add a way for listen_reverse to inform the caller when a new connection arrives. With the normal SSH listener, application code is called as new sessions are opened via the session_factory/process_factory callables provided to listen(), but nothing happens when the connection is opened unless you subclass SSHServer and implement the auth_completed() callback. In this case, though, I need the application to be called when a connection is opened so that it can actually start whatever sessions or other actions it needs to do when the connection completes, as all of that is initiated by the SSH client.

One thought I had was to implement an accept() method on the reverse listener. This would block until a new connection was received, returning the SSHClientConnection as soon as auth completed successfully or cleaning up the failed connection and raising an exception if a connection came in but auth didn't complete, much like how connect() works for outbound connections.

The caller in this case could sit in a loop calling accept(), starting a new async task for each successful reverse-direction connection that was returned that opened whatever sessions were needed, or set up port forwarding, or did whatever other operations were appropriate.

**Comment by silentnights** [comment]
Sorry for my very late response myself. I missed the notification from github.

I like the idea of accept as it looks much like a normal server code. But that also makes me wonder if I am thinking about it incorrectly. I mean I wonder if I am limiting my thinking to the old synchronous programming. !

I also realized that in my case I need to be informed once I actually get a higher level session connection, not just an ssh connection, so in my case I ended up passing a future to the protocol factory to pass it again to the session and set it's result there whether the connection is successful in terms of the higher layer session protocol or not. I await on this future and re-trigger a new connection if the previous one failed.

To be honest I hated that implementation. It looks ugly and passing a future all the was down the stream does not seem a correct way to do this, but I still have no better idea.

**Comment by silentnights** [comment]
Sorry for my very late response myself. I missed the notification from github.

I like the idea of accept as it looks much like a normal server code. But that also makes me wonder if I am thinking about it incorrectly. I mean I wonder if I am limiting my thinking to the old synchronous programming. !

I also realized that in my case I need to be informed once I actually get a higher level session connection, not just an ssh connection, so in my case I ended up passing a future to the protocol factory to pass it again to the session and set it's result there whether the connection is successful in terms of the higher layer session protocol or not. I await on this future and re-trigger a new connection if the previous one failed.

To be honest I hated that implementation. It looks ugly and passing a future all the was down the stream does not seem a correct way to do this, but I still have no better idea. Probably this can be done somehow using the protocol factory but I did not come up with a way to do it yet.

So the accept seems to be a better solution but if it can handle the case that the higher session protocol can set the future instead of doing it only on auth_completed then that would be a better solution.

But again I know my case is very special and for most of the usecases this won't be needed.

Again sorry for my late response.

**Comment by ronf** [comment]
No problem - this actually turned out not to be easy to crack. I started out with an implementation that involved a Future, but hid some of the details by providing a way to call a start() coroutine each time a new connection arrived that would perform the handshake and either return the new connection or raise an exception. However, that was very ugly. After that, I tried the accept() idea I mentioned, but maintaining a separate async queue for that and requiring a separate async task to read the queue, only to have to spawn new tasks in the application code for handling each connection and then properly cleaning up all those tasks when shutting things down was not pretty.

I ended up with an approach that allows you to pass in an "acceptor" and an "error_handler". The first is called on every successful handshake, passing you the new SSH connection and the latter is called on every failed handshake, passing you the failed connection (just to pull information like the address of the peer) and the exception raised during the handshake. This seems to be much cleaner, and lets you leave out one or both of these in cases when you don't need to do anything at the time a new connection arrives and only need to handle new sessions. Here's an example of what it looks like to open a listener for a reverse-direction client right now:

```python
async def run_client(conn):
    async with conn:
        result = await conn.run()
        print(result.stdout, end='')

async def listen_reverse():
    server = await asyncssh.listen_reverse(port=8022, known_hosts='known_hosts',
                                           acceptor=run_client)
                                               
    await server.wait_closed()

```

If you wanted to handle connection errors, you'd just pass an "error_handler" argument which could be either a callable or coroutine that takes "conn" and "exc" arguments.

Here's what it looks like to open a reverse-direction server that connects out:

```python
async def handle_client(process):
    process.stdout.write('Welcome to my SSH server!\n')
    process.exit(0)

async def connect_reverse():
    conn = await asyncssh.connect_reverse('localhost', 8022,
                                          server_host_keys=['ssh_host_key'],
                                          authorized_client_keys='ssh_user_ca',
                                          process_factory=handle_client)
    
    await conn.wait_closed()
```
This will support multiple incoming sessions on the single outbound connection, exiting only when the remote system closes the connection.

As part of the refactoring here, I have also expanded the connect_ssh() support I previously added to connect through jump hosts to also now support connect_reverse_ssh(), listen_ssh(), and listen_reverse_ssh() methods. So, in addition to supporting forward and reverse direction SSH connections directly over TCP, all of these cases can be supported tunneled over an already established SSH client connection with nearly identical syntax.

I'm still tracking down a minor issue related to reporting of address information in the case of "tunneled" connections, but as soon as I work that out I should have a first cut of this I can check into the "develop" branch.

**Comment by silentnights** [comment]
That looks pretty neat. I still do not get in "run_client" what "conn.run()" means, I assume you would want to establish some sort of a session with the server by calling conn.create_session().

But connect_reverse is exactly what I need. My current implementation is pretty ugly and using that new interface will make it much better.

Thanks a lot for the effort.

**Comment by ronf** [comment]
The conn.run() was a placeholder for opening an SSH channel and running some command within in. In this case, it launched an interactive shell and collected the output, but the server side just printed a welcome message and immediately closed the channel without accepting any input.

Since this is a reverse-direction connection, the request to open a channel actually comes from the side listening for new TCP connections, rather than the side opening the connection.

**Comment by ronf** [comment]
Ok - I've finally got a first cut for this reverse direction support in good enough shape to check in. It's now available in the develop branch as commit 0181d10 if you'd like to check it out!

**Comment by silentnights** [comment]
Ok, I will test it today and let you know how it goes.

**Comment by silentnights** [comment]
Sorry again for my late response. I have run into several issues in terms of dependencies and now I am getting Incompatible ssh peer (no acceptable kex algorithm) from my client which is using paramiko as an ssh backend.

I have seen that you are using a new version of cryptography, which I upgraded to. That let the application run, but I am not sure why I have such issue now, I am passing the same parameters to connect_reverse as in the create_server with the correct host and port for sure.

Any idea if anything is modified in this version? or if I am missing any parameters?

**Comment by ronf** [comment]
There shouldn’t be any changes to algorithms in this version. However, the reverse direction stuff can have an impact on the way known_hosts works, as the “client” that would normally validate against known_hosts wouldn’t only see an IP address from the “server” and not a hostname. Also, the port number would be a dynamic port, rather than the typical SSH port. This might cause the client to not send a proper list of trusted host keys, and that might lead to the error you saw around not finding a matching kex algorithm.

Do you have the debug output from the failure?

**Comment by silentnights** [comment]
I tried to enable the asyncssh logs using:

```
asyncssh.set_debug_level(3)
asyncssh.set_log_level(logging.DEBUG)
```

but still do not see any logs from asyncssh. I am doing that in my unittest.

But anyway, I found the issue. I was missing to pass server_host_keys to connect_reverse and later found that session_encoding is deprecated and encoding was used instead.

Now it all works. And I find the interface really good, this how I would assume to use it. I pass the same SSHServer as the one I pass to create_server so that the same server handles both types of connection.

I have one more question though. Now we create a connection to the client via connect_reverse. I will get no errors as long as the SSH connection is successful, now the client request a subsystem from the server and if the session created by the subsystem failed for some reason, the underlying SSH connection will still remain valid. Right?

I am trying to clarify that because in my use case I should maintain a connection with the client, but if I have a successful ssh connection to the client and the subsystem session fails, I would assume its the client responsibility to re-trigger the connection, but if the subsystem failure trigger a connection drop as well, then I will have to re-trigger a new connection.

Again thanks a lot for your help.

**Comment by ronf** [comment]
> I tried to enable the asyncssh logs using:
> 
>     asyncssh.set_debug_level(3)
>     asyncssh.set_log_level(logging.DEBUG)
>     but still do not see any logs from asyncssh. I am doing that in my unittest.

AsyncSSH’s logging depends on the standard Python “logging” module, and you’ll generally need to configure that before you use it. Generally, this is done with a call to logging.basicConfig(). The only argument you really need to give it there is the level, unless you want to also specify a file to output to. By default, it’ll output to stdout once this is called.

> But anyway, I found the issue. I was missing to pass server_host_keys to connect_reverse and later found that session_encoding is deprecated and encoding was used instead.

Yeah - you need to provide the server_host_keys, just as you would for create_server(). The only exception is if you wanted to use GSS (Kerberos) auth for key exchange. As for encoding, there’s still support for the older session_encoding/session_errors argument names in create_server(), but with the new function I decided to only support the preferred argument names of encoding and errors, which is what is used everywhere else.

> Now it all works. And I find the interface really good, this how I would assume to use it. I pass the same SSHServer as the one I pass to create_server so that the same server handles both types of connection.

Yeah, you can. I’m glad to hear you like the interface!

> I have one more question though. Now we create a connection to the client via connect_reverse. I will get no errors as long as the SSH connection is successful, now the client request a subsystem from the server and if the session created by the subsystem failed for some reason, the underlying SSH connection will still remain valid. Right?
> 
> I am trying to clarify that because in my use case I should maintain a connection with the client, but if I have a successful ssh connection to the client and the subsystem session fails, I would assume its the client responsibility to re-trigger the connection, but if the subsystem failure trigger a connection drop as well, then I will have to re-trigger a new connection.

Yes, that’s right. Opening of sessions will always be initiated by the listener when using connect_reverse/listen_reverse. So, once a connection is established by connect_reverse, that side doesn’t have to worry about reopening anything on session failures, but it would need to worry about connection failures. As you said, it might also need to handle connection closes that might be triggered by other errors.

**Comment by silentnights** [comment]
Thanks for the clarifications. Are you planning a release soon with this changes? I am looking forward to ditch my dirty workaround and use the new version. :-)

**Comment by ronf** [comment]
I was hoping to check it in over the weekend, but some other reports came in that I wanted to get to the bottom of before I cut a new release. If all goes well, I should have a 1.17 release out with this by next weekend, if not sooner.

**Comment by ronf** [comment]
This feature is now available in release 1.17.0!

**Comment by silentnights** [comment]
Thanks a lot. :-)
